src/libxtp/eigencuda.cc

```diff
@@ -0,0 +1,191 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <votca/xtp/eigencuda.h>
+
+namespace votca {
+namespace xtp {
+
+/*
+ * Deallocate memory from the device
+ */
+void free_mem_in_gpu(double *x) { checkCuda(cudaFree(x)); };
+
+EigenCuda::~EigenCuda() {
+
+  // destroy handle
+  cublasDestroy(_handle);
+  // destroy stream
+  cudaStreamDestroy(_stream);
+}
+
+/*
+ * Check if the available memory is enough to compute the system
+ */
+void EigenCuda::check_available_memory_in_gpu(size_t requested_memory) const {
+  // Use Unified memory
+  size_t *free, *total;
+  cudaMallocManaged(&free, sizeof(size_t));
+  cudaMallocManaged(&total, sizeof(size_t));
+  checkCuda(cudaMemGetInfo(free, total));
+
+  std::ostringstream oss;
+  oss << "There were requested : " << requested_memory
+      << "bytes int the device\n";
+  oss << "Device Free memory (bytes): " << *free
+      << "\nDevice total Memory (bytes): " << *total << "\n";
+
+  // Raise an error if there is not enough total or free memory in the device
+  if (requested_memory > *free) {
+    oss << "There is not enough memory in the Device!\n";
+    throw std::runtime_error(oss.str());
+  }
+
+  // Release memory
+  cudaFree(free);
+  cudaFree(total);
+}
+
+/*
+ * Allocate memory in the device for matrix A.
+ */
+uniq_double EigenCuda::alloc_matrix_in_gpu(size_t size_matrix) const {
+
+  // Pointer in the device
+  double *dmatrix;
+  checkCuda(cudaMalloc(&dmatrix, size_matrix));
+  uniq_double dev_ptr(dmatrix, free_mem_in_gpu);
+  return dev_ptr;
+}
+
+uniq_double EigenCuda::copy_matrix_to_gpu(const Mat &matrix) const {
+  // allocate memory in the device
+  size_t size_matrix = matrix.size() * sizeof(double);
+  uniq_double dev_ptr = alloc_matrix_in_gpu(size_matrix);
+
+  // Transfer data to the GPU
+  const double *hmatrix = matrix.data();  // Pointers at the host
+  cudaError_t err = cudaMemcpyAsync(dev_ptr.get(), hmatrix, size_matrix,
+                                    cudaMemcpyHostToDevice, _stream);
+  if (err != 0) {
+    throw std::runtime_error("Error copy arrays to device");
+  }
+  return dev_ptr;
+}
+
+/*
+ * Call the gemm function from cublas, resulting in the multiplication of the
+ * two matrices.
+ */
+void EigenCuda::gemm(const CudaMatrix &A, const CudaMatrix &B,
+                     CudaMatrix &C) const {
+
+  // Scalar constanst for calling blas
+  double alpha = 1.;
+  double beta = 0.;
+  const double *palpha = &alpha;
+  const double *pbeta = &beta;
+
+  cublasDgemm(_handle, CUBLAS_OP_N, CUBLAS_OP_N, A.rows(), B.cols(), A.cols(),
+              palpha, A.ptr(), A.rows(), B.ptr(), B.rows(), pbeta, C.ptr(),
+              C.rows());
+}
+
+/*
+ * \brief Perform a Tensor3D matrix multiplication
+ */
+std::vector<Mat> EigenCuda::right_matrix_tensor_mult(
+    const std::vector<Mat> &tensor, const Mat &B) const {
```

Comment:
this has to be void(std::vector<Mat> &tensor, const Mat &B), we cannot make a copy of tensor to return a result as it would double the required memory.

Reply:
done